tensorflow/c/kernels_experimental.cc

```diff
@@ -391,3 +396,217 @@ bool TF_IsRefInput(TF_OpKernelContext* ctx, int i, TF_Status* status) {
+    } else if (a.TypeId() == tensorflow::TypeIndex::Make<
+                                 ::tensorflow::data::OptionalVariant>()) {
+      TF_RETURN_IF_ERROR(
+          ValidateVariantType<::tensorflow::data::OptionalVariant>(a));
+      *out = ::tensorflow::data::OptionalVariant();
+
+      return ::tensorflow::data::OptionalBinaryAdd(
+          cc_ctx, *a.get<::tensorflow::data::OptionalVariant>(),
+          *b.get<::tensorflow::data::OptionalVariant>(),
+          out->get<::tensorflow::data::OptionalVariant>(), cc_binary_add_func);
+    }
+
+    const std::string type_index_name =
+        ::tensorflow::port::MaybeAbiDemangle(a.TypeId().name());
+
+    return ::tensorflow::errors::Internal(
+        "No unary variant binary_op function found for op ADD Variant "
+        "type_name: ",
+        type_index_name, " for device type: ", cc_ctx->device()->name());
+  };
+  ::tensorflow::AddNVariant(cc_ctx, binary_add_variant);
+  ::tensorflow::Set_TF_Status_from_Status(status, cc_ctx->status());
+}
+
+static Status ZerosLikeVariant(::tensorflow::OpKernelContext* cc_ctx,
+                               const Variant& input, Variant* out,
+                               void (*zeros_like_func)(TF_OpKernelContext* ctx,
+                                                       TF_Tensor* input,
+                                                       TF_Tensor* out)) {
+  auto cc_zeros_like_func = [zeros_like_func](
+                                ::tensorflow::OpKernelContext* cc_ctx,
+                                const Tensor& cc_input, Tensor* cc_out) {
+    AllocatorAttributes attr;
+    if (cc_input.dtype() == ::tensorflow::DT_VARIANT) {
+      attr.set_on_host(true);
+    }
+    TF_RETURN_IF_ERROR(cc_ctx->allocate_temp(cc_input.dtype(), cc_input.shape(),
+                                             cc_out, attr));
+
+    switch (cc_input.dtype()) {
+      case ::tensorflow::DT_INVALID: {
+        *cc_out = Tensor(::tensorflow::DT_INVALID);
+        break;
+      }
+      case ::tensorflow::DT_VARIANT: {
+        // If the wrapped tensor is also a variant, recursively call
+        // ZerosLikeVariant to unwrap it the same way
+        Variant* out_variant = cc_out->scalar<Variant>().data();
+        TF_RETURN_IF_ERROR(ZerosLikeVariant(cc_ctx,
+                                            cc_input.scalar<Variant>()(),
+                                            out_variant, zeros_like_func));
+        break;
+      }
+      default: {
+        Status status;
+        TF_Tensor* input = TF_TensorFromTensor(cc_input, &status);
+        TF_RETURN_IF_ERROR(status);
+
+        TF_Tensor* out = TF_TensorFromTensor(*cc_out, &status);
+        if (!status.ok()) {
+          TF_DeleteTensor(input);
+          return status;
+        }
+
+        auto* ctx = reinterpret_cast<TF_OpKernelContext*>(cc_ctx);
+        zeros_like_func(ctx, input, out);
+      }
+    }
+    return cc_ctx->status();
+  };
+
+  DCHECK_NE(out, nullptr);
```

Comment:
Can you replace all checks (debug or not) with status returns, please?

We have an internal presubmit checker that no longer allows checks due to the fact that is causes downtime / offers a chance for denial of service.

Debug checks are slightly worse as they are only active during debug builds but we rarely have them.

Reply:
@mihaimaruseac Sorry for the delay, I replaced the `CHECKs` with status returns.